chatkit-core/src/main/kotlin/com/pusher/chatkit/rooms/state/JoinedRoomsStateDiffer.kt

```diff
@@ -0,0 +1,65 @@
+package com.pusher.chatkit.rooms.state
+
+import com.pusher.chatkit.state.Action
+import com.pusher.chatkit.state.LeftRoom
+import com.pusher.chatkit.state.ReconnectJoinedRoom
+import com.pusher.chatkit.state.RoomUpdated
+import com.pusher.chatkit.state.State
+import org.reduxkotlin.GetState
+
+internal class JoinedRoomsStateDiffer(private val stateGetter: GetState<State>) {
+
+    fun stateExists() = stateGetter().joinedRoomsState != null
+
+    fun toActions(
+        newRooms: List<JoinedRoomInternalType>,
+        newUnreadCounts: Map<String, Int>
+    ): List<Action> =
+        joinedRoomActions(newRooms, newUnreadCounts) +
+            roomUpdatedActions(newRooms) +
+            leftRoomActions(newRooms)
+
+    private val currentRooms get() = stateGetter().joinedRoomsState!!.rooms
+
+    private fun joinedRoomActions(
+        newRooms: List<JoinedRoomInternalType>,
+        newUnreadCounts: Map<String, Int>
+    ): List<Action> =
+        newRooms.filter { !currentRooms.contains(it.id) }
+            .map { ReconnectJoinedRoom(it, newUnreadCounts[it.id]) }.toList()
+
+    private fun roomUpdatedActions(
+        newRooms: List<JoinedRoomInternalType>
+    ): List<Action> {
+        val actions = mutableListOf<Action>()
```

Comment:
can we apply this [feedback](https://github.com/pusher/chatkit-android/pull/193#discussion_r388862840) for all the functions in this file please?

Reply:
👍 i *think* i've made it better - i'm not very confident on functional programming so maybe you'd prefer to take a look over what I've done.

Reply:
looks very nice to me, thanks Danielle :)